Add support on Windows to install atexit handler(s) early in bootprocess.

lateralusX opened the pull request:
Add support on Windows to install atexit handler(s) early in application’s startup.

This is done by adding a new argument to install specific named custom handlers, --handlers=. There is also an argument to describe the supported handlers on current platform, --help-handlers.

The supported handlers can be different depending on implementing platform and currently only Windows uses this feature to install an atexit handler waiting for a keypress before exiting process. This is however a generic concept that could be used to run handler code early in the boot process on any platform for other purposes. Adding the argument first in command line makes it possible to do the installation very early in the boot process, this is used by the waitkeypress atexit handler in order to be used in cases where the arguments itself are unsupported or incorrect formatted causing an exit of the process.

When running under Visual Studio debugger the console window will not be kept open when the debugee process terminates. Normally this is not a problem, but there are scenarios, especially when having errors in configuration, when you would like the console to be kept open so you can read the output in order to detect the error.

Normally there is an option to start the console application from within Visual Studio but not attaching the debugger (CTRL+F5) and by default Visual Studio will then inject a pause in the execution of the process holding up the console, but for some reasons, having more complex solutions like mono, this feature stops working, so it is not possible to keep the console open in that scenario for more complex solutions.

This makes it a little problematic to hold the process when working within the debugger in order to look for console output and errors. Since mono can exit the process using exit command and not only by returning from main, the developer needs to set breakpoints on both end of main and in c-runtime exit methods. Even if that is a possible workaround this option won't work when running the process outside the debugger (CTRL+F5).

This commit adds a handler argument that can be used on Windows (potential on other platforms as well), atexit-waitkeypress.

The waitkeypress will hold the console and process open until a key is pressed and is very useful when running and debugging mono from within Visual Studio. Example of output when running mono’s mono-mini-regression-test with an assembly not found together with –handlers=atexit-waitkeypress command line argument:

failed to load assembly: gshared-nofound.exe
Overall results: tests: 0, 100% pass, opt combinations: 25
Press any key to continue . . .

Currently this is only implemented for Windows, but could be enabled for other platforms, if similar need exist elsewhere.

Since this is a standard mono command line option it is possible to inject it using the MONO_ENV_OPTIONS environment variable as well.

kumpera commented:
The code looks overly complicated.

There's no reason for it to be conditional.
This sort of code should be under the MONO_DEBUG env var.
The windows code should go to mini-windows.c

Here's my suggestion.

Add a wait_for_exit entry to debug_options.
Rename mono_runtime_install_handlers to mono_runtime_init_platform.
Move the #ifndef MONO_CROSS_COMPILE logic to init signals from mini-runtime.c into the actual functions.
In the mini-windows.c version of mono_runtime_init_platform check for wait_for_exit and install the at_exit handler.
Document it on the man page.

kumpera commented:
And once it's merged, document on the release notes.

lateralusX commented:
Thanks for the feedback! First let me give you some background on why the code looks like it currently does. One of the reason why I placed this code in main.c very early in the process init was because I wanted the exit handlers to be installed as early as possible in order to be able to catch more cases. One common problem when running from within the debugger is invalid configuration, like incorrect command line parameters. Hooking up all the logic into the existing infrastructure could risk getting the exit handler installed too late, missing out on a number of errors that could occur before the handler gets installed. As an example, mono_runtime_install_handlers won't be reached if you have an unknown command line parameter (or some other error in command line parameters), I assume there are other config errors that are detected before mono_runtime_install_handlers gets called that could be "missed".

The code installing the exit handler is data/table driven, that's one of the reasons why it could look a little more complicated then what to expect.  It also includes logic to consume the argument (if present) so other parts of the runtime won't need to care or handle it, since it is already "consumed". I also added the capabilities to always install it as a define option, so you don't need to add the startup argument to all existing configurations in order to use it during development/debugging.

I have no problem hooking up the logic into existing debug options and move the code around as you propose above, but then there will be some common scenarios when you still need to place breakpoints at end of main and in c-runtime exit methods, and then the value of having the "waitkeypress" exit handler will be reduced.

So given this background and additional information, what do you think?

vargaz commented:
It does look very complicated. Also, instead of the MONO_INSTALL_DEFAULT_ATEXIT_COMMANDS variable, it can use the existing MONO_ENV_OPTIONS variable, ie.
export MONO_ENV_OPTIONS="--atexit ..."

lateralusX commented:
I will try to simplify it by reducing the use cases a little and move code into windows-mini.c. I would like to keep the early install in one way or another in order to cover the use case of invalid command line arguments. This is the plan:

* Remove conditional option to always install exit handlers using define and use MONO_ENV_OPTION for similar feature (still needs to change debug configurations to set environment variables when launching from within VS, so not equivalent, but probably good enough).
* Remove logic to consume argument and teach runtime parsing logic to handle/ignore it (simplifies the implementation).
* Hook up "atexit" arguments with MONO_DEBUG env var or similar command line features.
* Move code into windows-mini.c.
* Keep early call to install atexit handlers to cover use case of invalid arguments. Need to look a little deeper into this to see if the parsing logic of mini_parse_debug_option mono_main could work, but then the first command line item needs to be the "atexit" option and the install needs to be done inside mini_parse_debug_option in order to be activated before any other arguments to be parsed.

I will go back to this PR once I finished another task currently in progress.

Thanks for all feedback!

kumpera commented:
Sounds great :)

lateralusX commented:
OK, finally I had some time to revisit this. I followed most of the plan above ending up with the following:

* Removed all code from main.c and moved parts into mini-runtime.c and mini-windows.c.
* Added a new argument, --handlers=HANDLERS used to install custom handlers during startup. This argument can be implemented differently on different platforms and currently only Windows implements one handler, atexit-waitkeypress. The implementation is done in the platform specific source file, mini-windows.c for Windows and for the rest "not implemented" in mini-runtime.c.
* By adding the argument first to command line or first in MONO_ENV_OPTIONS it is possible to install the handler very early in the boot process, enough to solve the use case of invalid arguments.
* Added argument to describe what handlers that are supported, --help-handlers. This is implemented by each platform, so each platform can have different support and help texts.
* Added description of --handlers argument to --help argument.

Since this is not only a debug feature I decided to go with a regular argument instead of just extending debug option.

It is possible to extend this feature in the future per platform. I already have a couple of other cases that needs to do things early in the boot process similar to the atexit-waitkeypress for none desktop Windows platforms.

lateralusX commented:
Change OK to merge?

vargaz commented:
Still looks a little bit too complicated, but its ok to merge.

lateralusX commented:
Aligned PR with WINAPI family support.
